File: src/apps/chifra/cmd/serve/names.go
package serve

/*-------------------------------------------------------------------------------------------
 * qblocks - fast, easily-accessible, fully-decentralized data from blockchains
 * copyright (c) 2016, 2021 TrueBlocks, LLC (http://trueblocks.io)
 *
 * This program is free software: you may redistribute it and/or modify it under the terms
 * of the GNU General Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version. This program is
 * distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program. If not, see http://www.gnu.org/licenses/.
 *-------------------------------------------------------------------------------------------*/

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

func EditName(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()
	newName := NewNamedAddress()
	for k := range r.Form {
		json.Unmarshal([]byte(k), &newName)
	}
	log.Print(Yellow, "Adding name: ", newName.ToJson(), Off)

	// Do the actual call
	cmd := exec.Command(GetCommandPath("ethNames"), "--create")
	cmd.Env = append(os.Environ(), "TB_NAME_ADDRESS="+newName.Address)
	cmd.Env = append(cmd.Env, "TB_NAME_NAME="+newName.Name)
	cmd.Env = append(cmd.Env, "TB_NAME_TAG="+newName.Tags)
	cmd.Env = append(cmd.Env, "TB_NAME_SOURCE="+newName.Source)
	cmd.Env = append(cmd.Env, "TB_NAME_SYMBOL="+newName.Symbol)
	cmd.Env = append(cmd.Env, "TB_NAME_DECIMALS="+newName.Decimals)
	cmd.Env = append(cmd.Env, "TB_NAME_DESCR="+newName.Description)
	out, err := cmd.Output()
	if err != nil {
		log.Print("Error from server: ", err)
	}
	output := string(out[:])
	log.Print(Yellow, string(output), Off)

	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept")
	w.Header().Set("Access-Control-Allow-Methods", "PUT, POST, GET, DELETE, OPTIONS")
	if strings.Contains(output, "\"errors\":") {
		w.WriteHeader(http.StatusBadRequest)
	} else {
		w.WriteHeader(http.StatusOK)
	}
	fmt.Fprintf(w, "{ \"data\": [ "+newName.ToJson()+" ]}")
}

type NamedAddress struct {
	Tags        string `json:"tags"`
	Address     string `json:"address"`
	Name        string `json:"name"`
	Symbol      string `json:"symbol,omitempty"`
	Source      string `json:"source"`
	Decimals    string `json:"decimals,omitempty"`
	Description string `json:"description,omitempty"`
	Deleted     bool   `json:"deleted,omitempty"`
	IsCustom    bool   `json:"is_custom,omitempty"`
	IsPrefund   bool   `json:"is_prefund,omitempty"`
	IsContract  bool   `json:"is_contract,omitempty"`
	IsErc20     bool   `json:"is_erc20,omitempty"`
	IsErc721    bool   `json:"is_erc721,omitempty"`
}

func NewNamedAddress() NamedAddress {
	ret := new(NamedAddress)
	ret.Name = "none"
	ret.Address = "none"
	ret.Tags = "19-Unknown"
	ret.Source = "EtherScan.io"
	// ret.Description = ""
	return *ret
}

func (name *NamedAddress) ToJson() string {
	e, err := json.Marshal(name)
	if err != nil {
		fmt.Printf("%s", err)
		return ""
	}
	return string(e)
}

File: src/other/names/names.cpp
/*-------------------------------------------------------------------------------------------
 * qblocks - fast, easily-accessible, fully-decentralized data from blockchains
 * copyright (c) 2016, 2021 TrueBlocks, LLC (http://trueblocks.io)
 *
 * This program is free software: you may redistribute it and/or modify it under the terms
 * of the GNU General Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version. This program is
 * distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program. If not, see http://www.gnu.org/licenses/.
 *-------------------------------------------------------------------------------------------*/
#include "etherlib.h"

namespace qblocks {
//-----------------------------------------------------------------------
class NameOnDisc {
  public:
    char tags[30 + 1];
    char address[42 + 1];
    char name[120 + 1];
    char symbol[30 + 1];
    char source[180 + 1];
    char description[255 + 1];
    uint16_t decimals;
    uint16_t flags;
    NameOnDisc(void);
    bool disc_2_Name(CAccountName& nm) const;
    bool name_2_Disc(const CAccountName& nm);
    string_q Format(void) const;
};
}  // namespace qblocks

#define NTESTS 20
#define NRUNS 500

//-----------------------------------------------------------------------
bool showName(CAccountName& name, void* data) {
    if (name.address == "0xf503017d7baf7fbc0fff7492b751025c6a78179b")
        return true;
    cout << name.Format(STR_DISPLAY_ACCOUNTNAME) << endl;
    return true;
}

//-----------------------------------------------------------------------
bool showName2025(NameOnDisc* name, void* data) {
    if (name && string_q(name->address) == "0xf503017d7baf7fbc0fff7492b751025c6a78179b")
        return true;
    cout << (name ? name->Format() : "") << endl;
    return true;
}

//-----------------------------------------------------------------------
CStringArray tests = {
    "0x43c65d1234edde9c9bc638f1fb284e1eb0c7ca1d",
    "0xa6a840e50bcaa50da017b91a0d86b8b2d41156ee",
    "0x29e240cfd7946ba20895a7a02edb25c210f9f324",
};

//-----------------------------------------------------------------------
int main(int argc, const char* argv[]) {
    etherlib_init(quickQuitHandler);

    for (size_t x = 0; x < NTESTS; x++) {
        if (argc < 2) {
            loadNames(true);
            forEveryName(true, showName, nullptr);
            for (size_t i = 0; i < NRUNS; i++) {
                for (auto test : tests) {
                    CAccountName name;
                    findName(true, test, name);
                    cout << name << endl;
                    findToken(true, test, name);
                    cout << name << endl;
                }
            }
            clearNames(true);
            cerr << "A-" << x << endl;

        } else if (argc < 3) {
            loadNames(false);
            forEveryName(false, (NAMEFUNC)showName2025, nullptr);
            for (size_t i = 0; i < NRUNS; i++) {
                for (auto test : tests) {
                    CAccountName name;
                    findName(false, test, name);
                    cout << name << endl;
                    findToken(false, test, name);
                    cout << name << endl;
                }
            }
            clearNames(false);
            cerr << "B-" << x << endl;
        }
    }

    etherlib_cleanup();
    return 1;
}
